### DISCUSSION TITLE
Documentation for ClpSimplex::createPiecewiseLinearCosts() API function?

### DISCUSSION THREAD

**Discussion by Antarctic-Fox** [coin-or/Clp#213] [Q&A]
I want to reformulate my LP problem in CLP solver using a convex piecewise linear objective (really just need an absolute value without creating extra linear constraints for it). API function createPiecewiseLinearCosts looks like a good fit but it is very poorly documented. I have checked the headers [1], the user guide [2], googled for examples online [3]. None describe the parameters of the function. **If the original developer is still on the team, is there any hope to get a clear description of the parameters expected by the function?**
--Thanks.

[1] Description from the ClpSimplex header does not describe parameters:
     /** Constructs a non linear cost from list of non-linearities (columns only)
         First lower of each column is taken as real lower
         Last lower is taken as real upper and cost ignored

         Returns nonzero if bad data e.g. lowers not monotonic
     */
     int createPiecewiseLinearCosts(const int * starts,
                                    const double * lower, const double * gradient);

Description from ClpNonLinearCost.hpp makes no sense either: 
     /** Constructor from simplex and list of non-linearities (columns only)
         First lower of each column has to match real lower
         Last lower has to be <= upper (if == then cost ignored)
         This could obviously be changed to make more user friendly
     */
     ClpNonLinearCost(ClpSimplex * model, const int * starts,
                      const double * lower, const double * cost);

[2] The user guide only mentions it in passing:
The primal can also use a non linear cost which should work for piecewise linear convex functions. 

[3] Best online example I've found is on this website:
https://cpp.hotexamples.com/examples/-/ClpSimplex/-/cpp-clpsimplex-class-examples.html

The code is very convoluted with no comments and no explanations, it mingles a bunch of unrelated stuff into the usage of createPiecewiseLinearCosts().

**Comment by jjhforrest** [comment]
Piecewise linear seems to work - but can print incorrect objective value.

Should be able to fix soon in master.

Will put up a better example.

John Forrest

**Comment by Antarctic-Fox** [comment]
Thank you, John. A simple example to demo this feature would be great.
